### ISSUE TITLE
NodeError: Cannot call write after a stream was destroyed in VSCode Debug

### ISSUE THREAD

**Issue by sangeeth96** [expressjs/morgan#209]
I am using VSCode's `launch.json` with debug mode and debug console along with nodemon to restart the server after file changes. The app is a simple express application for a REST API. I am using morgan just as is for logging HTTP request/response.

To make my life easier, I've added the following to my `launch.json` to start the app and reconnect the debugger automatically since nodemon will kill the server on file changes. On the very first run of my application via VSCode, things work just fine. I get logs on debug console as expected and no problems. However, as soon as I change a file and nodemon kicks in and VSCode debugger disconnects and connects again, when I hit the API, I get the following:

```
Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at doWrite (_stream_writable.js:411:19)
    at writeOrBuffer (_stream_writable.js:399:5)
    at Socket.Writable.write (_stream_writable.js:299:11)
    at logRequest (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/morgan/index.js:130:14)
    at logger (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/morgan/index.js:135:7)
    at Layer.handle [as handle_request] (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:317:13)
    at /Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:275:10)
    at expressInit (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/middleware/init.js:40:5)
    at Layer.handle [as handle_request] (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:317:13)
    at /Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:275:10)
    at query (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/middleware/query.js:45:5)
    at Layer.handle [as handle_request] (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:317:13)
    at /Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:275:10)
    at Function.handle (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/router/index.js:174:3)
    at Function.handle (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/application.js:174:10)
    at Server.app (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/express/lib/express.js:39:9)
    at Server.emit (events.js:198:13)
    at parserOnIncoming (_http_server.js:677:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
```

Or sometimes, it is this one which causes the server to crash:

```
NodeError: Cannot call write after a stream was destroyed
    at doWrite (_stream_writable.js:411:19)
    at writeOrBuffer (_stream_writable.js:399:5)
    at Socket.Writable.write (_stream_writable.js:299:11)
    at Array.logRequest (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/morgan/index.js:130:14)
    at listener (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/on-finished/index.js:169:15)
    at onFinish (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/on-finished/index.js:100:5)
    at callback (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/ee-first/index.js:55:10)
    at ServerResponse.onevent (/Users/ssudheer/Develop/greenqloud/fabric-cvs-provisioning-service/node_modules/ee-first/index.js:93:5)
    at ServerResponse.emit (events.js:203:15)
    at onFinish (_http_outgoing.js:671:10)
    at process._tickCallback (internal/process/next_tick.js:61:11)
Emitted 'error' event at:
    at errorOrDestroy (internal/streams/destroy.js:107:12)
    at onwriteError (_stream_writable.js:430:5)
    at onwrite (_stream_writable.js:461:5)
    at doWrite (_stream_writable.js:411:11)
    at writeOrBuffer (_stream_writable.js:399:5)
    [... lines matching original stack trace ...]
    at process._tickCallback (internal/process/next_tick.js:61:11)
```

I thought this was a bug with morgan initially, updated it to the latest stable release and tried again but I kept getting these errors. On further research, I noticed that this **only happens within VSCode's debugger context** and not when I am directly running it in the CLI. That too, only after a debugger reconnect and not on the very first app start. Can someone shine a light on if this is a VSCode specific issue that's causing this to morgan (I didn't experience this with anything else) or am I doing something wrong?

**Comment by sangeeth96** [comment]
Okay, so I discovered a setting for launch configuration items, `"console": "integratedTerminal"` which switched the output to the integrated terminal. With this setting, I'm not seeing any issues which means this could very well be associated with VSCode and not morgan. If someone thinks so too, Imma move this to microsoft/code.

**Comment by sappusaketh** [comment]
I had same issue as yours. Able to resolve by updating `launch.json`. In your `launch.json` check runtime version is same as your global node version.

**Comment by dougwilson** [comment]
So based on the stack trace, it looks as if `process.stdout` stream was destroyed at some point. This module could be updated such that it doesn't try to write to a destroyed stream, which would mean you don't get the error any longer, but it seems like it would likely just result is no further morgan output once that happens. I'll have to install and play around with VSCode to see what is going on unless someone knows of a solution.

**Comment by ryhinchey** [comment]
@sangeeth96 there's also a setting called `outputCapture` which can be set to `std` which will capture logs that write to `stdout` like morgan does by default.  The issue you were experiencing was definitely with vscode and is resolved via settings available in vscode.

I'm going to go ahead and close this issue but don't hesitate to open a new one if logs aren't writing as expected in the future!

**Comment by kopax** [comment]
Hi, I am having the same issue that started to appear within a verdaccio docker  container and I do not understand it, do you guys have any hint? This became critical since it break all our CIs

**Comment by ryhinchey** [comment]
hi @kopax do you mind opening a new issue with more specifics around what you're doing with verdaccio? This issue was created for vscode specifically so it'd be great to get a new issue specific for the issue you're having with verdaccio. Please add a minimal reproducible example as well. Thanks!

**Comment by tgirotto** [comment]
it seems to be indeed a problem with visual studio code. By adding the property "console" (see below) the problem disappears.

```
{
            "name": "Nodemon",
            "type": "node",
            "request": "launch",
            "cwd": "${workspaceFolder}",
            "runtimeExecutable": "nodemon",
            "runtimeArgs": [
                "--inspect-brk"
            ],
            "program": "${workspaceFolder}/bin/www",
            "restart": true,
            "console": "integratedTerminal",
            "internalConsoleOptions": "neverOpen"
        }
```

**Comment by bersuc** [comment]
For me it happened because i switched the VSCode default terminal from WSL to powershell. Returning it to the previous one, it worked.
